### DISCUSSION TITLE
`type-safe-api` migration questions?

### DISCUSSION THREAD

**Discussion by zetashift** [aws/aws-pdk#356] [Q&A] [answered]
I didn't check up on this repo and I just saw that `open-api-gateway` is going to be replaced by something nicer. I would like to change my code(now that it's still not in production :P), to use type-safe-api, but  `aws-prototyping-sdk/type-safe-api` can't be found:
```
.projenrc.ts:6:36 - error TS2307: Cannot find module 'aws-prototyping-sdk/type-safe-api' or its corresponding type declarations.

6 import { TypeSafeApiProject } from "aws-prototyping-sdk/type-safe-api"
```
Same happens if I use the `@` package:

```
.projenrc.ts:6:36 - error TS2307: Cannot find module '@aws-prototyping-sdk/type-safe-api' or its corresponding type declarations.

6 import { TypeSafeApiProject } from "@aws-prototyping-sdk/type-safe-api"
```
It does say that `SmithyApiGatewayTsProject` is deprecated? I ran `npx projen upgrade` so I should be on the latest version.
Or am I just too soon with upgrading?

**Comment by cogwirrel** [comment, marked as answer]
It is published! :) You will need to add a dependency on the new package though as well as upgrading the other packages :) eg if you're using the `nx-monorepo`:

```ts
const monorepo = new NxMonorepoProject({
  name: "my-project",
  defaultReleaseBranch: "main",
  devDeps: ["aws-prototyping-sdk", "@aws-prototyping-sdk/type-safe-api"],
});
```

Also note that because it's a dependency that the `.projenrc.ts` makes use of, you need to do 2 passes with projen:

* Add the dependency to the monorepo project, then `npx projen` to add the dependency
* Now the dependency is added, you can import `TypeSafeApiProject` from `@aws-prototyping-sdk/type-safe-api` in your `.projenrc.ts`, configure it, and `npx projen` again

Hope that helps! :)

**Reply by zetashift** [reply]
Ugh, can't believe I missed that! Thank you :D

**Reply by cogwirrel** [reply]
No worries!! Hope the migration goes smoothly! Let me know if you have any issues :)

**Comment by zetashift** [comment]
@cogwirrel the migration so far is going good. I have a question tho about where the code of the Lambda handlers/routers should go: Should I create a `src` directory manually in my `api` directory, should I put Lambda handlers in the `runtime` folder and then define the CDK construct in `api/infrastructure/typescript/`?

**Reply by cogwirrel** [reply]
Nice! What I'd recommend is creating a separate infrastructure project (eg with `new AwsCdkTypeScriptApp` or `new PDKPipelineTsProject` in your `.projenrc.ts`), and instantiating the `Api` construct in that new project. I'd also recommend putting the handlers in that same new infrastructure project if you're using `NodejsFunction`, or if you want more separation you can create a separate `TypeScriptProject` for the handlers, and have their build task create the bundle with esbuild.

It's best not to add code to `api/infrastructure/*` or `ap/runtime/*` as the source folders are all `.gitignored` and generated as part of build :)

Hope that helps! I notice the quick start guide isn't very clear where things should live so I'll look at updating that soon! :)

**Reply by zetashift** [reply]
I'll make  a new TypeScriptProject for it! 

~~One more double-check, I had to delete the Nx cache to get the old projects out of the build, but is it true that for `type-safe-api` that is named `api`, the following builds will be run by Nx:~~ ( I found out it does a few things based off Projen projects).

```
    - api-typescript-infra
    - api-typescript-runtime
    - api-model
    - api
```

**Comment by zetashift** [comment]
I finished the migration, some notes about the process:

- One might have to do `npx nx clear-cache` to let Nx pickup the new build tasks/remove the old ones
- `pnpm build` messed up generating the typescript-runtime project. Doing `yarn build` or `npx nx run-many ..` works fine. (I kinda wish it was only pnpm, not the biggest fan of yarn v1)
- I wasn't smart enough to figure out what the `esbuild` build task for the `TypeScriptProject` for the handlers was. So I now I only use the `infra` project. I renamed it `platform` because it does some business logic as well as infra.

Not related to PDK:
- My frontend projen config started to fail to build. It used SolidJS/solid-start. In the end I switched to SvelteKit, because that just worked in 1 go. I don't think any tooling can keep up with the frontend JS world :P.

All in all, looking good! Thank you for all the help and I like PDK a lot :D.

**Reply by cogwirrel** [reply]
Awesome glad you made it! 😄

Ahh what was the issue with `pnpm build` when generating the `typescript-runtime` project? If that doesn't work it's definitely worth raising a bug as we hope to support pnpm, yarn and npm :)

Ah cool, that works! In case it's helpful, something like this for a standalone TypeScriptProject for lambdas might do the trick:

```ts
const lambdas = new TypeScriptProject({
  parent: monorepo,
  outdir: 'packages/lambdas',
  ...
  deps: [
    api.runtime.typescript!.package.packageName,
  ],
  devDeps: [
    'esbuild',
  ],
});

// Build the lambda dist
lambdas.packageTask.reset('mkdir -p lambdadist && esbuild src/index.ts --bundle --platform=node --outfile=lambdadist/index.js');
lambdas.gitignore.addPatterns("lambdadist");
```

Then in your infrastructure, (assuming you export a function for every api handler, eg `export const sayHello = sayHelloHandler(...)`) you can do something like:

```ts
const fn = new Function(this, 'Lambda', {
    handler: "index.sayHello",
    runtime: Runtime.NODEJS_18_X,
    code: Code.fromAsset('path/to/lambdadist'),
  });
```

Hope that makes sense! :)

**Reply by zetashift** [reply]
> Ahh what was the issue with pnpm build when generating the typescript-runtime project? If that doesn't work it's definitely worth raising a bug as we hope to support pnpm, yarn and npm :)

I couldn't reproduce it! So I didn't make an issue. Deleting the `node_modules` fixed things, so probably I just did something funky like `yarn install` and then `pnpm build` and the `api-model:generate` task got in some weird state.

Thank you! `'mkdir -p lambdadist && esbuild src/index.ts --bundle --platform=node --outfile=lambdadist/index.js'` was what I was looking for!

**Reply by cogwirrel** [reply]
Ahh gotcha! I've experienced some funkiness with switching from `yarn` to `pnpm` if I don't nuke `node_modules` first too - so sounds like it could easily be something like that :) Thanks!

Great to hear! 😄

**Comment by zetashift** [comment]
I've been trying to get everything over using `pnpm` but I'm still having trouble. For the one at $WORK I can't reproduce yet, but when I try to setup a `nx-monorepo` project using `pnpm install` I also get a failure:

## Steps to reproduce

```sh
npx projen new --from aws-prototyping-sdk nx-monorepo && pnpm install
```

Error message: 
```
~/d/w/pnpm-repro-fail 15.2s ❱ pnpm projen                                                                                                                                     devenv-shell-env

> pnpm-repro-fail@0.0.0 projen /home/rishi/dev/webdev/pnpm-repro-fail
> npx projen

👾 default | ts-node --project tsconfig.dev.json .projenrc.ts

/home/rishi/dev/webdev/pnpm-repro-fail/node_modules/.pnpm/parse-json@2.2.0/node_modules/parse-json/index.js:33
                throw jsonErr;
                ^
ErrorEXError [JSONError]: Unexpected token '/' at 1:1 in /home/rishi/dev/webdev/pnpm-repro-fail/tsconfig.json
// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
^
// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
^
    at module.exports (/home/rishi/dev/webdev/pnpm-repro-fail/node_modules/.pnpm/parse-json@2.2.0/node_modules/parse-json/index.js:27:17)
    at parseFileSync (/home/rishi/dev/webdev/pnpm-repro-fail/node_modules/.pnpm/tsconfig@1.1.0/node_modules/tsconfig/dist/tsconfig.js:91:16)
    at Object.readFileSync (/home/rishi/dev/webdev/pnpm-repro-fail/node_modules/.pnpm/tsconfig@1.1.0/node_modules/tsconfig/dist/tsconfig.js:75:12)
    at readConfig (/home/rishi/dev/webdev/pnpm-repro-fail/node_modules/.pnpm/ts-node@0.2.0/node_modules/ts-node/dist/ts-node.js:19:38)
    at Object.register (/home/rishi/dev/webdev/pnpm-repro-fail/node_modules/.pnpm/ts-node@0.2.0/node_modules/ts-node/dist/ts-node.js:41:18)
    at Object.<anonymous> (/home/rishi/dev/webdev/pnpm-repro-fail/node_modules/.pnpm/ts-node@0.2.0/node_modules/ts-node/dist/bin/ts-node.js:33:25)
    at Module._compile (node:internal/modules/cjs/loader:1267:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1321:10)
    at Module.load (node:internal/modules/cjs/loader:1125:32)
    at Module._load (node:internal/modules/cjs/loader:965:12) {
  fileName: '/home/rishi/dev/webdev/pnpm-repro-fail/tsconfig.json'
}
```

So do I _need_ to do `yarn install` on the first install always? Because that works fine.

**Reply by cogwirrel** [reply]
Looks like you're doing things right! (Note you can add `--packageManager pnpm` to the `npx projen new` command to avoid having to delete node_modules etc.)

It looks like that specific error you're encountering could be due to `ts-node` somehow being resolved to a super old version `ts-node@0.2.0` which possibly didn't understand comments in `tsconfig.json` back then? :)

I was able to reproduce this too - I think the problem is that `projen` uses `*` for dependencies that are specified without versions, and `pnpm`'s default behaviour (contrary to `yarn` and `npm`) is to resolve the lowest version possible! I actually opened an issue for this with projen yesterday: https://github.com/projen/projen/issues/2628

As a workaround until that issue is fixed, you can add this to your `.projenrc.ts`:

```ts
const npmrc = new NpmConfig(monorepo);
npmrc.addConfig('resolution-mode', 'highest');
```

You'll need to manually create this file (`.npmrc` in the root dir) initially too to avoid the chicken-egg problem of already having old versions and not being able to run `projen` to create it:

```
resolution-mode=highest
```

Then, manually reset all the versions of dependencies/devDependencies in your `package.json` to `"*"`, delete `node_modules` and `pnpm-lock.yaml`, and run `pnpm i` and `pnpm projen` - after that you should see the `package.json` includes newer versions of everything and all is good!

**Reply by zetashift** [reply]
Not sure if this is an issue from me or not but(seems some ESLint issue?):

When running:
```
pnpx projen new --from aws-prototyping-sdk --packageManager pnpm nx-monorepo
```
Getting the following error:
```

> micp@0.0.0 eslint
> pnpm exec projen eslint


 >  NX   tasksRunner is not a function

   Pass --verbose to see the stacktrace.

Error: Command failed: npm run eslint --if-present
👾 eslint | pnpm exec nx run-many --target=eslint --output-style=stream --nx-bail
👾 Task "eslint" failed when executing "pnpm exec nx run-many --target=eslint --output-style=stream --nx-bail" (cwd: /home/rishi/dev/micp)

    at checkExecSyncError (node:child_process:885:11)
    at Object.execSync (node:child_process:957:15)
    at exec (/home/rishi/.local/share/pnpm/store/v3/tmp/dlx-45755/node_modules/.pnpm/projen@0.71.31/node_modules/projen/lib/util.js:15:19)
    at initProject (/home/rishi/.local/share/pnpm/store/v3/tmp/dlx-45755/node_modules/.pnpm/projen@0.71.31/node_modules/projen/lib/cli/cmds/new.js:325:25)
    at initProjectFromModule (/home/rishi/.local/share/pnpm/store/v3/tmp/dlx-45755/node_modules/.pnpm/projen@0.71.31/node_modules/projen/lib/cli/cmds/new.js:279:11)
    at Object.handler (/home/rishi/.local/share/pnpm/store/v3/tmp/dlx-45755/node_modules/.pnpm/projen@0.71.31/node_modules/projen/lib/cli/cmds/new.js:94:16)
    at Object.runCommand (/home/rishi/.local/share/pnpm/store/v3/tmp/dlx-45755/node_modules/.pnpm/projen@0.71.31/node_modules/projen/node_modules/yargs/build/index.cjs:446:48)
    at Object.parseArgs [as _parseArgs] (/home/rishi/.local/share/pnpm/store/v3/tmp/dlx-45755/node_modules/.pnpm/projen@0.71.31/node_modules/projen/node_modules/yargs/build/index.cjs:2706:53)
    at Object.runCommand (/home/rishi/.local/share/pnpm/store/v3/tmp/dlx-45755/node_modules/.pnpm/projen@0.71.31/node_modules/projen/node_modules/yargs/build/index.cjs:410:36) {
  status: 1,
  signal: null,
  output: [
    null,
    null,
    <Buffer f0 9f 91 be 20 1b 5b 33 37 6d 1b 5b 34 6d 65 73 6c 69 6e 74 1b 5b 32 34 6d 20 7c 20 70 6e 70 6d 20 65 78 65 63 20 6e 78 20 72 75 6e 2d 6d 61 6e 79 20 ... 204 more bytes>
  ],
  pid: 45910,
  stdout: null,
  stderr: <Buffer f0 9f 91 be 20 1b 5b 33 37 6d 1b 5b 34 6d 65 73 6c 69 6e 74 1b 5b 32 34 6d 20 7c 20 70 6e 70 6d 20 65 78 65 63 20 6e 78 20 72 75 6e 2d 6d 61 6e 79 20 ... 204 more bytes>
}
 ERROR  Command failed with exit code 1: projen new --from aws-prototyping-sdk --packageManager pnpm nx-monorepo

pnpm: Command failed with exit code 1: projen new --from aws-prototyping-sdk --packageManager pnpm nx-monorepo
    at makeError (/nix/store/vzv3w7nbn71s13i6kxdzsh24ldphcpmk-pnpm-8.3.1/lib/node_modules/pnpm/dist/pnpm.cjs:24230:17)
    at handlePromise (/nix/store/vzv3w7nbn71s13i6kxdzsh24ldphcpmk-pnpm-8.3.1/lib/node_modules/pnpm/dist/pnpm.cjs:24801:33)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Object.handler [as dlx] (/nix/store/vzv3w7nbn71s13i6kxdzsh24ldphcpmk-pnpm-8.3.1/lib/node_modules/pnpm/dist/pnpm.cjs:206754:7)
    at async /nix/store/vzv3w7nbn71s13i6kxdzsh24ldphcpmk-pnpm-8.3.1/lib/node_modules/pnpm/dist/pnpm.cjs:216086:21
    at async main (/nix/store/vzv3w7nbn71s13i6kxdzsh24ldphcpmk-pnpm-8.3.1/lib/node_modules/pnpm/dist/pnpm.cjs:216053:34)
    at async runPnpm (/nix/store/vzv3w7nbn71s13i6kxdzsh24ldphcpmk-pnpm-8.3.1/lib/node_modules/pnpm/dist/pnpm.cjs:216308:5)
    at async /nix/store/vzv3w7nbn71s13i6kxdzsh24ldphcpmk-pnpm-8.3.1/lib/node_modules/pnpm/dist/pnpm.cjs:216300:7

**Reply by cogwirrel** [reply]
Hmm I think there are a few issues related to the uber package `aws-prototyping-sdk`, so the preferred way is to create a new project using `@aws-prototyping-sdk/nx-monorepo` instead.

Here's the up and coming PR to deprecate the uber package: https://github.com/aws/aws-prototyping-sdk/pull/384

And here's the one which updates the docs: https://github.com/aws/aws-prototyping-sdk/pull/371 :)

**Reply by zetashift** [reply]
I've migrated my repo to the `@aws-prototyping-sdk/nx-monorepo` project yesterday because of the discussion thread. 

Waiting for the `npmrc` release/fix to migrate to PNPM! But all in all, I've written a parts of `db <-> API <-> frontend` application with PDK, and I never really done that before, so even with yarn it's all good.

**Reply by cogwirrel** [reply]
Ah nice :) The npmrc fix (https://github.com/aws/aws-prototyping-sdk/pull/373) should hopefully be released soon! I just hit retry on the release as there was a transient issue :)

**Reply by cogwirrel** [reply]
Ok it should be there in 0.17.5! https://github.com/aws/aws-prototyping-sdk/releases/tag/v0.17.5

**Reply by zetashift** [reply]
I've switched over to `pnpm` and the npmrc error is indeed gone, but no dev depencies are being installed in the node_modules folders.

One of my subprojects has the following devdeps in its `package.json`:
```
  "devDependencies": {
    "@sveltejs/adapter-auto": "^2.0.1",
    "@sveltejs/kit": "^1.15.9",
    "@types/node": "^16",
    "autoprefixer": "^10.4.14",
    "npm-check-updates": "^16",
    "postcss": "^8.4.23",
    "projen": "^0.71.23",
    "svelte": "^3.58.0",
    "svelte-check": "^3.2.0",
    "tailwindcss": "^3.3.2",
    "tslib": "^2.5.0",
    "typescript": "^4.9.5",
    "vite": "^4.3.3"
  },
```

The `dev` script is `vite dev` so when I run that I get the following error:

```
~/micx main• ❱ pnpm exec nx run dashboard:dev

> nx run dashboard:dev

> dashboard@0.0.0 dev /Users/rishad/micx/packages/dashboard
> npx projen dev
👾 dev | vite dev
/bin/sh: vite: command not found
👾 Task "dev" failed when executing "vite dev" (cwd: /Users/rishad/micx/packages/dashboard)
 ELIFECYCLE  Command failed with exit code 1.

 ——————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Running target dev for project dashboard failed

   Failed tasks:

   - dashboard:dev

   Hint: run the command with --verbose for more details.
```

I've tried wiping all the node_modules and running `pnpm install` and `pnpm projen` but still the same error. Not sure if this is on me

**Reply by cogwirrel** [reply]
Hmmm how do you define your script for vite dev in your `.projenrc.ts`? Is it via creating a task eg. `project.addTask('dev').exec('vite dev')`? Or do you use `setScript`? If you're doing the former it should work... if not it could be a bug!

A potential workaround might be to change your task from `vite dev` to `pnpm exec vite dev`? :)

**Reply by zetashift** [reply]
It's the former:

```
    const devTask = this.tasks.addTask("dev");
    this.tasks.removeTask("build");
    const buildTask = this.tasks.addTask("build");
    devTask.exec("vite dev");
    buildTask.exec("vite build");
```

Even `pnpm exec vite dev` fails, I think the problem is that dev dependencies aren't being installed at all.
When I look in the `node_modules` folder there is no vite folder to be seen. 

I'll try to get a repro going :P.

**Reply by cogwirrel** [reply]
Very odd! Nice that'd be great if you could raise an issue with a minimal example :)

**Comment by zetashift** [comment]
Also (hopefully my last question :P), how do I change the name of the `DefaultApi` client that's created? I would like to have something like `FooApi` instead of it, but I'm not sure what knob to change in `ts-client` config of the smithy projections config.

**Reply by cogwirrel** [reply]
Haha no shame in lots of questions! The more questions (and hopefully answers!) the more it helps other users too! 😄 

You can change the name of the API client by using tags on your operations, eg:

```smithy
@tags(["Foo"])
operation DoSomething {
  ...
}
```

Will result in the `DoSomething` operation being included in a class called `FooApi` :) The generated clients will group operations into separate classes based on their _first_ tag - so you could even end up with separate `FooApi` and `BarApi` classes if you want to separate your operations. Operations that don't have any tags get put in `DefaultApi`. It's worth noting that any subsequent tags (second, third etc) are just used for grouping operations in generated documentation.

I'll usually use the same first tag for all my operations to collate them in the same client class, but then add extra tags to group things nicely in the generated `DocumentationFormat.HTML_REDOC` docs :) eg. some operations might have `@tags(["Store", "Pets"])` and others might have `@tags(["Store", "Payments"])` :)

**Reply by zetashift** [reply]
Crystal clear! Thank you for all the help, I'm learning Smithy, PDK, projen as I go :P, so this all very helpful (also sorry I noticed the docs mention this as well).

**Comment by zetashift** [comment]
- PNPM migration works!
- DB generator project works!

I had 30 minutes of bliss programming, when I did an `pnpm projen upgrade` and borked my `type-safe-api`. There seems to be a subtle version mismatch between the `constructs` library:

```
 src/apiIntegration.ts(21,7): error TS2322: Type 'OperationConfig<{ integration: LambdaIntegration; }>' is not assignable to type 'ApiIntegrations'.
mic-platform:   Type '{ integration: LambdaIntegration; }' is not assignable to type 'TypeSafeApiIntegration'.
mic-platform:     The types of 'integration.render' are incompatible between these types.
mic-platform:       Type '(props: import("/home/rishi/dev/work/micx/node_modules/.pnpm/@aws-prototyping-sdk+type-safe-api@0.18.2_aws-cdk-lib@2.78.0_cdk-nag@2.26.14_constructs@10.2.15_projen@0.71.38/node_modules/@aws-prototyping-sdk/type-safe-api/lib/construct/integrations/integrations").IntegrationRenderProps) => import("/home/rishi/dev/work...' is not assignable to type '(props: import("/home/rishi/dev/work/micx/node_modules/.pnpm/@aws-prototyping-sdk+type-safe-api@0.17.5_aws-cdk-lib@2.77.0_cdk-nag@2.26.13_constructs@10.2.14_projen@0.71.37/node_modules/@aws-prototyping-sdk/type-safe-api/lib/construct/integrations/integrations").IntegrationRenderProps) => import("/home/rishi/dev/work...'.
mic-platform:         Types of parameters 'props' and 'props' are incompatible.
mic-platform:           Type 'import("/home/rishi/dev/work/micx/node_modules/.pnpm/@aws-prototyping-sdk+type-safe-api@0.17.5_aws-cdk-lib@2.77.0_cdk-nag@2.26.13_constructs@10.2.14_projen@0.71.37/node_modules/@aws-prototyping-sdk/type-safe-api/lib/construct/integrations/integrations").IntegrationRenderProps' is not assignable to type 'import("/home/rishi/dev/work/micx/node_modules/.pnpm/@aws-prototyping-sdk+type-safe-api@0.18.2_aws-cdk-lib@2.78.0_cdk-nag@2.26.14_constructs@10.2.15_projen@0.71.38/node_modules/@aws-prototyping-sdk/type-safe-api/lib/construct/integrations/integrations").IntegrationRenderProps'.
mic-platform:             The types of 'scope.node' are incompatible between these types.
mic-platform:               Type 'import("/home/rishi/dev/work/micx/node_modules/.pnpm/constructs@10.2.14/node_modules/constructs/lib/construct").Node' is not assignable to type 'import("/home/rishi/dev/work/micx/node_modules/.pnpm/constructs@10.2.15/node_modules/constructs/lib/construct").Node'.
mic-platform:                 Types have separate declarations of a private property 'host'.
```

This error seems eerily similiar to this: https://github.com/aws/aws-cdk/issues/3416 so I don't think it's a PDK issue, but can I somehow pin/resolve versions for `constructs` and `cdk` in the monorepo?

**Reply by cogwirrel** [reply]
Hey! Ah yeah that happens fairly regularly when upgrading cdk/constructs versions! The best option is usually to add package resolutions. You can do something like:

```ts
monorepo.package.addPackageResolutions("constructs@xxx", "aws-cdk-lib@yyy");
```

That should hopefully pin your constructs/cdk versions across all packages in the workspace :)

**Reply by zetashift** [reply]
Sweet, now I can get back to blissful programming!

Thank you :D
